### PR TITLE
Replicate aliases in cross-cluster replication

### PR DESCRIPTION
**Pull request by jasontedor**
This commit adds functionality so that aliases that are manipulated on leader indices are replicated by the shard follow tasks to the follower indices. Note that we ignore write indices. This is due to the fact that follower indices do not receive direct writes so the concept is not useful.

Relates #41396

### CODE REVIEW

File: x-pack/plugin/ilm/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/indexlifecycle/CCRIndexLifecycleIT.java [hunk 1]
```diff
@@ -281,6 +291,74 @@ public void testCcrAndIlmWithRollover() throws Exception {
...
+            // Wait for the index to continue with its lifecycle and be shrunk
+            assertBusy(() -> assertTrue(indexExists(shrunkenIndexName)));
+
+            // assert the aliases were replicated
+            assertBusy(() -> {
+                for (int i = 0; i < numberOfAliases; i++) {
+                    assertOK(client().performRequest(new Request("HEAD", "/" + shrunkenIndexName + "/_alias/alias_" + i)));
+                }
+            });
+            assertOK(client().performRequest(new Request("HEAD", "/" + shrunkenIndexName + "/_alias/" + indexName)));
```
Comment (jasontedor) [inline x-pack/plugin/ilm/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/indexlifecycle/CCRIndexLifecycleIT.java]:
```suggestion
            assertBusy(() -> assertOK(client().performRequest(new Request("HEAD", "/" + shrunkenIndexName + "/_alias/" + indexName))));
```

File: x-pack/plugin/ilm/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/indexlifecycle/CCRIndexLifecycleIT.java [hunk 2]
```diff
@@ -108,6 +108,15 @@ public void testBasicCCRAndILMIntegration() throws Exception {
         }
     }
 
+    public void test() throws IOException {
```
Comment (gwbrown) [inline x-pack/plugin/ilm/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/indexlifecycle/CCRIndexLifecycleIT.java]: This doesn't appear to be used anywhere and should be removed.

If I'm wrong and it is used, it should have a more descriptive name.

**Review by AthenaEryma — APPROVED** [review]
LGTM with one very minor exception (see comment). 

The CCR/ILM integration tests cover the scenarios we discussed. I'm not an expert on CCR, although I didn't see any obvious problems in the code or tests.

One thing that I don't necessarily want to block on but did notice is that there doesn't seem to be a way to opt out of replicating aliases.  I can see that possibly being annoying or surprising to users, for example in the use case where you want "bidirectional" replication between two clusters, for example:
- Cluster A has `myindex-cluster-a`, which is written to by an application locally and replicated to Cluster B
- Cluster B has `myindex-cluster-b`, which is replicated to Cluster A
- The user has an alias `myindex` pointing to both indices set up on both A and B, but differing in that each has `is_write_index` set on the local index.

This would allow the user to both read from and write to `myindex` on both clusters, which would be convenient from an application perspective, but will break upon upgrading to a version with these changes.

To be clear, I'm not suggesting that we try to come up with a solution that allows this setup to coexist with replicated aliases, just that this change might break existing setups and it might be worth allowing users to opt out of alias replication entirely (at least until 8.0), either per-follower or for the whole cluster.

**Comment by jasontedor** [comment]
@elasticmachine update branch

**Review by martijnvg — APPROVED** [review]
LGTM

**Pull request merged**